pkg/msgs/can_msgs.go

```diff
@@ -1,17 +1,158 @@
 package msgs
 
 import (
+	"encoding/json"
+	"io/ioutil"
+	"sort"
+	"strconv"
 	"time"
+
+	log "github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	canpb "telemetry/pkg/protos"
 )
 
 // CAN describes a single message from the CAN bus.
 type CAN struct {
-	ID        uint32 `json:"id"`
-	Timestamp uint64 `json:"timestamp"`
-	Data      uint64 `json:"data"`
+	Source    uint8                  `json:"source"`
+	ID        uint16                 `json:"id"`
+	RTR       bool                   `json:"rtr"`
+	Timestamp uint64                 `json:"timestamp"`
+	Data      map[string]interface{} `json:"data"`
+}
+
+// ToSlice converts a CAN object to a list of string with the Data field as a JSON object.
+func (c CAN) ToSlice() []string {
+	mapBytes, _ := json.Marshal(c.Data)
+	return []string{
+		strconv.FormatUint(uint64(c.Source), 10),
+		strconv.FormatUint(uint64(c.ID), 10),
+		strconv.FormatBool(c.RTR),
+		strconv.FormatUint(c.Timestamp, 10),
+		string(mapBytes),
+	}
+}
+
+var schema canpb.CanSchema
+
+// CanMsgInit Initializes the CAN msg parser with the provided schema.
+func CanMsgInit(filename string) error {
+	b, err := ioutil.ReadFile(filename)
+	if err != nil {
+		log.Fatalln("Failed to read can_schema.asciipb")
+		return err
+	}
+	err = proto.UnmarshalText(string(b), &schema)
+	if err != nil {
+		log.Fatalln("Failed to unmarshal:", err.Error())
+		return err
+	}
+	return nil
 }
 
 // NewCAN creates a CAN struct, this should be used for all incoming messages.
-func NewCAN(id uint32, data uint64) CAN {
-	return CAN{id, uint64(time.Now().UnixNano()) / uint64(time.Millisecond), data}
+func NewCAN(rawID uint32, rawData uint64) CAN {
+	var canMsg CAN
+	canMsg.Data = make(map[string]interface{})
+	parse(rawID, rawData, &canMsg)
+	return canMsg
+}
+
+func parse(rawID uint32, rawData uint64, canMsg *CAN) {
+	canMsg.Timestamp = uint64(time.Now().UnixNano()) / uint64(time.Millisecond)
+	canMsg.Source = uint8(rawID & 0xF)
+	canMsg.RTR = ((rawID & 0x10) >> 4) == 1
+	canMsg.ID = uint16((rawID & 0x7E0) >> 5)
+	srcMsg := findCanMessage(uint32(canMsg.ID))
+	if srcMsg == nil {
+		log.Errorf("Error: no matching schema for", canMsg.ID)
+		canMsg.Data["raw"] = rawData
+	}
+	switch srcMsg.GetCanData().GetFrame().(type) {
+	case *canpb.CanData_U8:
+		data := srcMsg.GetCanData().GetU8()
+		name := data.GetFieldName_1()
```

Comment:
Too bad this can't be an array or something

Reply:
Yeah, but such is the nature of Go. That's why there is often a lot of codegen done for Go.